pymc_experimental/tests/model/test_marginal_model.py

```diff
@@ -251,6 +268,125 @@ def test_marginalized_change_point_model_sampling(disaster_model):
     )
 
 
+def test_recover_marginals_basic():
+    with MarginalModel() as m:
+        sigma = pm.HalfNormal("sigma")
+        p = np.array([0.5, 0.2, 0.3])
+        k = pm.Categorical("k", p=p)
+        mu = np.array([-3.0, 0.0, 3.0])
+        mu_ = pt.as_tensor_variable(mu)
+        y = pm.Normal("y", mu=mu_[k], sigma=sigma)
+
+    m.marginalize([k])
+
+    rng = np.random.default_rng(211)
+
+    with m:
+        prior = pm.sample_prior_predictive(
+            samples=20,
+            random_seed=rng,
+            return_inferencedata=False,
+        )
+        idata = InferenceData(posterior=dict_to_dataset(prior))
+
+    idata = m.recover_marginals(idata, return_samples=True)
+    assert "k" in idata
+    assert "lp_k" in idata
+    assert idata.k.shape == idata.y.shape
+    assert idata.lp_k.shape == idata.k.shape + (len(p),)
+
+    def true_logp(y, sigma):
+        y = y.repeat(len(p)).reshape(len(y), -1)
+        sigma = sigma.repeat(len(p)).reshape(len(sigma), -1)
+        return log_softmax(
+            np.log(p)
+            + norm.logpdf(y, loc=mu, scale=sigma)
+            + halfnorm.logpdf(sigma)
+            + np.log(sigma),
+            axis=1,
+        )
+
+    np.testing.assert_almost_equal(
+        true_logp(idata.y.values.flatten(), idata.sigma.values.flatten()),
+        idata.lp_k[0].values,
+    )
+
+
+def test_recover_batched_marginal():
+    """Test that marginalization works for batched random variables"""
+    with MarginalModel() as m:
+        sigma = pm.HalfNormal("sigma")
+        idx = pm.Bernoulli("idx", p=0.7, shape=(2, 2))
```

Comment:
Here I would give a different length to each dim and check the come out correctly in the idata

Reply:
I nice feature (no need to block this PR, can be a follow up issue) would be to reuse dims in the computed variables if the user specified dims for the marginalized variables